core/src/data_sources/data_source.rs

```diff
@@ -1266,6 +1172,122 @@ impl DataSource {
         Ok(documents)
     }
 
+    async fn retrieve_chunks_without_query(
+        &self,
+        store: Box<dyn Store + Sync + Send>,
+        qdrant_client: &Arc<QdrantClient>,
+        top_k: usize,
+        filter: &Option<SearchFilter>,
+    ) -> Result<Vec<(String, Chunk)>> {
+        let store = store.clone();
+
+        let (doc_ids, _) = store
+            .find_data_source_document_ids(
+                &self.project,
+                self.data_source_id(),
+                filter,
+                // with top_k documents, we should be guaranteed to have at
+                // least top_k chunks, if we make the assumption that each
+                // document has at least one chunk.
+                Some((top_k, 0)),
+            )
+            .await?;
+
+        let qdrant_batch_size: usize = 10; // number of `document_ids` to query at once in QDrant
+        let qdrant_page_size: u32 = 100; // number of points to fetch per page in QDrant
+
+        let mut chunks: Vec<(String, Chunk)> = vec![];
+
+        // iterate over the doc_ids in batches of qdrant_batch_size
+        for batch in doc_ids.chunks(qdrant_batch_size) {
+            let document_id_hashes = batch
+                .iter()
+                .map(|document_id| {
+                    let mut hasher = blake3::Hasher::new();
+                    hasher.update(document_id.as_bytes());
+                    format!("{}", hasher.finalize().to_hex())
+                })
+                .collect::<Vec<_>>();
+            let document_id_condition = qdrant::FieldCondition {
+                key: "document_id_hash".to_string(),
+                r#match: Some(qdrant::Match {
+                    match_value: Some(qdrant::r#match::MatchValue::Keywords(
+                        qdrant::RepeatedStrings {
+                            strings: document_id_hashes,
+                        },
+                    )),
+                }),
+                ..Default::default()
+            }
+            .into();
+            let qdrant_batch_filter = qdrant::Filter {
+                must: vec![document_id_condition],
+                ..Default::default()
+            };
+
+            let mut page_offset: Option<PointId> = None;
+            let mut batch_points: Vec<RetrievedPoint> = vec![];
+
+            // we must scroll through all result pages of a batch, because we want to
+            // to sort by reverse-chron timestamp and then by chunk_offset
+            // and QDrant doesn't support any kind of sorting
+            loop {
```

Comment:
I get antsy with potential infinite looping, can we add a sentinel to defense ourselves against qdrant going loco?

Reply:
It's so ugly I agree, but infinite loops are worse?

Reply:
Yeah sounds good 👍

Reply:
as per IRL: went with 16k iterations max. If we do reach 16k iterations, then we'll return an error.